readme.md
#### Audio Overload SDK - Development Release 1.4.8  February 15, 2009

Copyright (c) 2007-2009 R. Belmont and Richard Bannister.
All rights reserved.


Please refer to `license.txt` for the specific licensing details of this
software.

----

This SDK opens up some of the music file format engines developed for the
Audio Overload project.
You may use this code to play the formats on systems we don't support or
inside of applications other than AO.

##### Compilation
Make sure that you've pulled all Git submodules together with this repo:
```
	git clone --recursive https://github.com/nmlgc/aosdk.git
```

On Windows, MinGW is required to build the code through the Makefile.
Cygwin's own gcc is unsupported, so make sure to specify MinGW's gcc in the
`CC` environment variable when building on Cygwin:

```
	# For 32-bit builds
	CC=i686-w64-mingw32-gcc make
	# For 64-bit builds
	CC=x86_64-w64-mingw32-gcc make
```

On Linux, set the environment variable `OSTYPE` to `linux` before you run
`make`:

```
	OSTYPE=linux make
```

To build on a big-endian platform, change `LSB_FIRST=1` to `=0` in the
Makefile.

By default, the code is compiled with optimizations using the `release` make
target. Use the `debug` target to generate debugging information and disable
optimizations:

```
	make debug
```

New in Release 1.4.8
- Guard against invalid data sometimes created by makessf.py (fixes crashing
  Pebble Beach ST-V rips)

----

##### MIDI dumping support

The following table lists the MIDI events generated by the MIDI dumper for
each individual engine, as well as the loss of data in this conversion.

| Engine                          |  Note On (Pitch) | Note On (Velocity) | Note Off |
|---------------------------------|:----------------:|:------------------:|:--------:|
| Capcom QSound (.qsf)            |                  |                    |          |
| Sega Saturn (.ssf)              |                  |                    |          |
| Sony PlayStation 1 (.psf, .spu) |                  |                    |          |
| Sega Dreamcast (.dsf)           | ✔ (- 5½ octaves) |     ✔ (- 1 bit)    |     ✔    |
| Sony PlayStation 2 (.psf2)      |                  |                    |          |

----

Entry points of an AO engine are as follows:

* `int32 XXX_start(uint8 *, uint32)`

	This function attempts to recognize and load a file of a specific
	type.  It is assumed external code has already checked the file's
	signature in cases where that's possible.  The first parameter is a
	pointer to the entire file in memory, and the second is the length of
	the file.  The return value is `AO_SUCCESS` if the engine properly
	loaded the file and `AO_FAIL` if it didn't.

* `int32 XXX_sample(stereo_sample_t *)`

	This function actually plays the song, generates the next sample in
	16-bit stereo format at 44100 Hz, and returns it through the given
	pointer.

* `int32 XXX_frame(void)`

	This function is called once per frame and can be used to update the
	internal state of the format engine if necessary.

* `int32 XXX_stop(void)`

	This function ceases playback and cleans up the engine.  You must call
	`_start()` again after this to play more music.

* `int32 XXX_command(int32, int32)`

	For some engines, this allows you to send commands while a song is
	playing.  The first parameter is the command (these are defined in
	`ao.h`), the second is the parameter.  These commands are as follows:

	* `COMMAND_PREV` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this moves back one song.

	* `COMMAND_NEXT` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this moves forward one song.

	* `COMMAND_RESTART` (parameter ignored) - Restarts the current song
	  from the beginning.  Not supported by all engines.

	* `COMMAND_HAS_PREV` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this checks if moving backwards
	  from the current song is a valid operation.  (Returns `AO_FAIL` if
	  not)

	* `COMMAND_HAS_NEXT` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this checks if moving forward
	  from the current song is a valid operation.  (Returns `AO_FAIL` if
	  not)

	* `COMMAND_GET_MIN` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this returns the lowest valid
	  song number.
	* `COMMAND_GET_MAX` (parameter ignored) - for file formats which have
	  more than one song in a file (NSF), this returns the highest valid
	  song number.

	* `COMAND_JUMP` - for file formats which have more than one song in a
	  file (NSF), this command jumps directly to a specific song number,
	  which is passed in as the parameter.

* `int32 XXX_fillinfo(ao_display_info *)`

	This function fills out the `ao_display_info` struct (see `ao.h` for
	details) with information about the currently playing song.  The
	information provided varies by engine.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### Added
- On Windows, it is now possible to specify a custom playback device on the
  command line using the new `-d/--device` option.  Additionally, all valid
  playback devices can be listed by passing `--list-devices`.
- Instrument samples are now dumped to .wav files for the following engines:
  * Dreamcast: low-level
- The program now terminates once the song has ended.  Useful for batch
  processing.

#### Changes to the Makefile:
- The Makefile should now detect 64-bit Linux systems automatically; however,
  specifying `OSTYPE=linux` is still necessary.
- The Makefile now comes with two build targets: a `release` target which
  compiles the code with optimizations, and a `debug` target that generates
  debugging information and disables optimizations.
- Audio Overload's own `CC`, `CPP`, `CFLAGS`, `LDFLAGS` and `LIBS` are now
  added to the environment variables of the same names during the build.
- Binaries are now suffixed with GCC's build machine identifier.

### Changed
- The README file now is formatted using Markdown syntax.
- Moved the Ikaruga sample into the samples/ subdirectory.
- Corlett tags now are stored using an unlimited key→value hash table and are
  no longer limited to 255 bytes.  In return, though, all `corlett_t`
  structures now need to be freed using `corlett_free()` in order to not leak
  the memory used for the tags.
- The `ao_display_info` structure now uses char pointers instead of fixed-size
  char buffers.

### Fixed
- On Windows, Unicode filenames passed on the command line are now supported
  correctly.
- Corlett length and fade tags with more than three fractional digits are now
  supported correctly.

### Removed
- Due to the inclusion of [win32_utf8](https://github.com/thpatch/win32_utf8)
  for transparent handling of Unicode file names on Windows, compilation on
  Cygwin's GCC is no longer supported.

## [1.4.8] - 2009-02-15
### Fixed
- Guard against invalid data sometimes created by `makessf.py` (fixes crashing
  Pebble Beach ST-V rips).

## [1.4.7] - 2008-11-03
### Fixed
- Fixed slightly too slow tempo for DSFs (shown by Virtua Fighter 4 DSF rip).

## [1.4.6] - 2008-11-02
### Fixed
- More AICA/DSF fixes for games using the DTPK driver.

## [1.4.5] - 2008-11-02
### Fixed
- AICA/DSF fixes for AEG readback, invalid loop parameters, and AICA Timer B
  operation.
- On newer Linux distros with PulseAudio `/dev/dsp` is missing but `/dev/dsp1`
  exists.  Try both.

## [1.4.4] - 208-10-19
### Fixed
- AICA/DSF fixes for Naomi rips using the full 8 MB of RAM.  (kingshriek)

## [1.4.3] - 2008-07-28
### Fixed
- More AICA fixes and cleanups.  (ajax16384)

## [1.4.2] - 2008-07-28
### Fixed
- Many AICA fixes.  Sound quality is improved and DC offset is fixed.  (Deunan
  Knute, ajax16384)

## [1.4.1] - 2008-04-23
### Fixed
- SCSP fix for crash cymbal issue in Shining the Holy Ark.  Use with
  kingshriek's new re-rip for best results.

## [1.4.0] - 2008-03-07
### Added
- DSF format engine, after some AICA fixes/improvements and some ARM7 cleanup.
  This is considered good enough to call a non-alpha.

## [1.3.1] - 2008-01-19
### Fixed
- Further FM and pitch improvements to SCSP. (kingshriek)

## [1.3] - 2008-01-13
### Added
- Added length and fade tag support to SSF. (R. Belmont)

### Fixed
- Fixed FM bug in SCSP for SSF files. (kingshriek)

## [1.2.1] - 2008-01-05
### Fixed
- Working FM synthesis in the SCSP, so more SSF files sound good.

## [1.2.0] - 2007-12-20
### Fixed
- Major overhaul of the PSF2 support.  More games work, including many
  Squaresoft titles.

## [1.1.7] - 2007-12-16
### Fixed
- kingshriek fixed an SCSP pitch calculation overflow and added FM support.

## [1.1.6] - 2007-12-14
### Fixed
- kingshriek strikes again: SCSP playback quality is much improved!

## [1.1.5] - 2007-12-12
### Fixed
- kingshriek strikes again: SCSP panning and other fixes, plus interpolation
  merged from MAME.

## [1.1.4] - 2007-12-11
### Fixed
- Even more SSF & SCSP fixes from kingshriek.

## [1.1.3] - 2007-12-09
### Fixed
- Even more SSF & SCSP fixes from kingshriek.

## [1.1.2] - 2007-12-02
### Fixed
- More SSF & SCSP fixes from kingshriek.

## [1.1.1] - 2007-12-01
### Fixed
- More SSF & SCSP fixes from kingshriek.

## [1.1] - 2007-11-06
### Fixed
- Two fixes for the SSF are included from kingshriek.

### Added
- PSF, PSF2, and SPU format engines are included.
- Known working example songs are included.


[unreleased]: https://github.com/nmlgc/aosdk/compare/v1.4.8...HEAD
[1.4.8]: https://github.com/nmlgc/aosdk/compare/v1.4.7...v1.4.8
[1.4.7]: https://github.com/nmlgc/aosdk/compare/v1.4.6...v1.4.7
[1.4.6]: https://github.com/nmlgc/aosdk/compare/v1.4.5...v1.4.6
[1.4.5]: https://github.com/nmlgc/aosdk/compare/v1.4.4...v1.4.5
[1.4.4]: https://github.com/nmlgc/aosdk/compare/v1.4.3...v1.4.4
[1.4.3]: https://github.com/nmlgc/aosdk/compare/v1.4.2...v1.4.3
[1.4.2]: https://github.com/nmlgc/aosdk/compare/v1.4.1...v1.4.2
[1.4.1]: https://github.com/nmlgc/aosdk/compare/v1.4.0...v1.4.1
[1.4.0]: https://github.com/nmlgc/aosdk/compare/v1.3.1...v1.4.0
[1.3.1]: https://github.com/nmlgc/aosdk/compare/v1.3...v1.3.1
[1.3]: https://github.com/nmlgc/aosdk/compare/v1.2.1...v1.3
[1.2.1]: https://github.com/nmlgc/aosdk/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/nmlgc/aosdk/compare/v1.1.7...v1.2.0
[1.1.7]: https://github.com/nmlgc/aosdk/compare/v1.1.6...v1.1.7
[1.1.6]: https://github.com/nmlgc/aosdk/compare/v1.1.5...v1.1.6
[1.1.5]: https://github.com/nmlgc/aosdk/compare/v1.1.4...v1.1.5
[1.1.4]: https://github.com/nmlgc/aosdk/compare/v1.1.3...v1.1.4
[1.1.3]: https://github.com/nmlgc/aosdk/compare/v1.1.2...v1.1.3
[1.1.2]: https://github.com/nmlgc/aosdk/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/nmlgc/aosdk/compare/v1.1...v1.1.1
[1.1]: https://github.com/nmlgc/aosdk/compare/v1.0...v1.1


ao.h
//
// Audio Overload SDK
//
// Fake ao.h to set up the general Audio Overload style environment
//

#ifndef __AO_H
#define __AO_H

#include <stdio.h>

#define AO_SUCCESS					1
#define AO_FAIL						0
#define AO_FAIL_DECOMPRESSION		-1

#define AUDIO_RATE					(44100)

enum
{
	COMMAND_NONE = 0,
	COMMAND_PREV,
	COMMAND_NEXT,
	COMMAND_RESTART,
	COMMAND_HAS_PREV,
	COMMAND_HAS_NEXT,
	COMMAND_GET_MIN,
	COMMAND_GET_MAX,
	COMMAND_JUMP
};

/* Compiler defines for Xcode */
#ifdef __BIG_ENDIAN__
	#undef LSB_FIRST
#endif

#ifdef __LITTLE_ENDIAN__
	#define LSB_FIRST	1
#endif

typedef unsigned char ao_bool;

#ifdef __GNUC__
#include <stddef.h>	// get NULL
#include <stdbool.h>

#ifndef TRUE
#define TRUE  (1)
#endif
#ifndef FALSE
#define FALSE (0)
#endif

#endif

#ifdef _MSC_VER
#include <stddef.h>	// get NULL

#ifndef TRUE
#define TRUE  (1)
#endif
#ifndef FALSE
#define FALSE (0)
#endif

#define true (1)
#define false (0)

#define strcasecmp _strcmpi

#endif

#ifndef PATH_MAX
#define PATH_MAX	2048
#endif

typedef struct
{
	const char *title[9];
	const char *info[9];
} ao_display_info;

typedef unsigned char		uint8;
typedef unsigned char		UINT8;
typedef signed char			int8;
typedef signed char			INT8;
typedef unsigned short		uint16;
typedef unsigned short		UINT16;
typedef signed short		int16;
typedef signed short		INT16;
typedef signed int			int32;
typedef unsigned int		uint32;
#ifdef LONG_IS_64BIT
typedef signed long             int64;
typedef unsigned long           uint64;
#else
typedef signed long long	int64;
typedef unsigned long long	uint64;
#endif

#ifdef WIN32
#include "win32_utf8/src/entry.h"

#ifndef _BASETSD_H
typedef signed int			INT32;
typedef unsigned int		UINT32;
typedef signed long long	INT64;
typedef unsigned long long	UINT64;
#endif
#else
typedef signed int			INT32;
typedef unsigned int		UINT32;
#ifdef LONG_IS_64BIT
typedef signed long         INT64;
typedef unsigned long       UINT64;
#else
typedef signed long long	INT64;
typedef unsigned long long	UINT64;
#endif
#endif

#ifndef INLINE
#if defined(_MSC_VER)
#define INLINE __forceinline
#elif defined(__GNUC__)
#define INLINE __inline__
#elif defined(_MWERKS_)
#define INLINE inline
#elif defined(__powerc)
#define INLINE inline
#else
#define INLINE
#endif
#endif

static uint16 INLINE SWAP16(uint16 x)
{
	return (
		((x & 0xFF00) >> 8) |
		((x & 0x00FF) << 8)
	);
}

static uint32 INLINE SWAP24(uint32 x)
{
	return (
		((x & 0xFF0000) >> 16) |
		((x & 0x00FF00) << 0) |
		((x & 0x0000FF) << 16)
	);
}

static uint32 INLINE SWAP32(uint32 x)
{
	return (
		((x & 0xFF000000) >> 24) |
		((x & 0x00FF0000) >> 8) |
		((x & 0x0000FF00) << 8) |
		((x & 0x000000FF) << 24)
	);
}

#if LSB_FIRST
#define BE16(x) (SWAP16(x))
#define BE24(x) (SWAP24(x))
#define BE32(x) (SWAP32(x))
#define LE16(x) (x)
#define LE24(x) (x)
#define LE32(x) (x)

#ifndef __ENDIAN__ /* Mac OS X Endian header has this function in it */
#define Endian32_Swap(x) (SWAP32(res))
#endif

#else

#define BE16(x) (x)
#define BE24(x) (x)
#define BE32(x) (x)
#define LE16(x) (SWAP16(x))
#define LE24(x) (SWAP24(x))
#define LE32(x) (SWAP32(x))

#endif

typedef struct {
	int16 l;
	int16 r;
} stereo_sample_t;

int ao_get_lib(const char *filename, uint8 **buffer, uint64 *length);

#endif // AO_H

extern volatile ao_bool ao_song_done;

/// Portability functions defined in ao.c
/// -------------------------------------
FILE* ao_fopen(const char *fn, const char *mode);

// We don't care about permissions, even on Linux
int ao_mkdir(const char *dirname);

#define fopen ERROR_use_ao_fopen_instead!
#define mkdir ERROR_Use_ao_mkdir_instead!
/// -------------------------------------

#include "wavedump.h"
#include "sampledump.h"
